burr/integrations/haystack.py

```diff
@@ -0,0 +1,197 @@
+        self._reads = list(reads.keys()) if isinstance(reads, dict) else reads
+        self._writes = list(writes.values()) if isinstance(writes, dict) else writes
+        self._bound_params = bound_params if bound_params is not None else {}
+
+        self._socket_mapping = {}
+        if isinstance(reads, dict):
+            for state_field, socket_name in reads.items():
+                self._socket_mapping[socket_name] = state_field
+
+        if isinstance(writes, dict):
+            for socket_name, state_field in writes.items():
+                self._socket_mapping[socket_name] = state_field
+
+    @property
+    def reads(self) -> list[str]:
+        return self._reads
+
+    @property
+    def writes(self) -> list[str]:
+        return self._writes
+
+    @property
+    def inputs(self) -> tuple[dict[str, str], dict[str, str]]:
+        """Return dictionaries of required and optional inputs."""
+        required_inputs, optional_inputs = {}, {}
+        for socket_name, input_socket in self._component.__haystack_input__._sockets_dict.items():
+            state_field_name = self._socket_mapping.get(socket_name, socket_name)
+
+            if state_field_name in self.reads:
+                continue
+            elif state_field_name in self._bound_params:
+                continue
+
+            if input_socket.default_value == haystack_empty:
+                required_inputs[state_field_name] = input_socket.type
+            else:
+                optional_inputs[state_field_name] = input_socket.type
+
+        return required_inputs, optional_inputs
+
+    def run(self, state: State, **run_kwargs) -> dict[str, Any]:
+        """Call the Haystack `Component.run()` method. It returns a dictionary
+        of results with mapping {socket_name: value}.
+
+        Values come from 3 sources:
+        - bound parameters (from HaystackAction instantiation, or by using `.bind()`)
+        - state (from previous actions)
+        - run_kwargs (inputs from `Application.run()`)
+        """
+        values = {}
+
+        # here, precedence matters. Alternatively, we could unpack all dictionaries at once
+        # which would throw an error for key collisions
+        for param, value in self._bound_params.items():
+            values[param] = value
+
+        for param in self.reads:
+            values[param] = state[param]
+
+        for param, value in run_kwargs.items():
+            values[param] = value
+
+        return self._component.run(**values)
+
+    def update(self, result: dict, state: State) -> State:
+        """Update the state using the results of `Component.run()`."""
+        state_update = {}
+        for socket_name, value in result.items():
+            state_field_name = self._socket_mapping.get(socket_name, socket_name)
+            if state_field_name in self.writes:
+                state_update[state_field_name] = value
+
+        return state.update(**state_update)
+
+    def bind(self, **kwargs):
```

Comment:
Otherwise we can make this a more specific name to not conflict

Reply:
If I understand, the pattern is to only allow `bound_params` at init?

This would be ok:
```python
haystack_retrieve_documents = HaystackAction(
    component=InMemoryEmbeddingRetriever(InMemoryDocumentStore()),
    name="retrieve_documents",
    reads=["query_embedding"],
    writes=["documents"],
    bound_params={"foo": "bar"},
)
```
This would not be ok?
```python
haystack_retrieve_documents = HaystackAction(
    component=InMemoryEmbeddingRetriever(InMemoryDocumentStore()),
    name="retrieve_documents",
    reads=["query_embedding"],
    writes=["documents"],
)
haystack_retrieve_documents.bind(foo="bar")
```

I should simply remove the `.bind()` method?

Reply:
I think so, or call it something else? Alternatively, we can make bind work at the action level. Currently it's only for functions, and class-based actions don't have it.

Reply:
So yeah, the pattern should be a copy-operation here, or remove this and just do the inputs in the parameters.